### PR TITLE
Language contributions: phase one

### LINKED ISSUE #74: Add translations for ALFONSÓ languages
[just for fun]
Add translation for

- [x] Latin
- [ ] French
- [x] Italian
- [ ] German
- [x] Spanish
- [ ] Ancient greek

For english continuous translation is required.

### LINKED ISSUE #162: Supportability for locales
Currently it's hard to follow the state of different locales, makes keeping them up to date even harder. We should somehow improve on this stage.
@uno20001, do you know of any extension that could help us?
If there is nothing, I think it would make sense to implement it as a page. This way, uran-users could translate the page themselves and we only need to review and commit them.

### LINKED ISSUE #265: Create FE and BE for localization page
The current UI is basic and an option for rewrite.

### PR DESCRIPTION
**Pull request by kdmnk**
Now every user can help translate the site.
See comments below.

Fixed #74 
Fixed #162 
Fixed #265

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -6,6 +6,12 @@
 /public/mix-manifest.json
 /storage/*.key
 /vendor
+/resources/lang/de
+/resources/lang/fr
+/resources/lang/gr
+/resources/lang/it
+/resources/lang/la
+/resources/lang/sp
```
Comment (kdmnk) [inline .gitignore:14]: Removed the language files that is not hungarian or english (users can not modify those).
The motivation was that changing the files on the server side could cause us problems, so we can manage the language and hungarian translations here, and the users can manage the rest based on our work, separately.
But these files have to be deleted from git to .gitignore take effect and that would erase our work so far (eg. latin translations).

File: app/Console/Commands/AddExpression.php [hunk 2]
```diff
@@ -48,7 +48,8 @@ public function handle()
         $expression_key = explode('.', $this->argument('key'))[1];
         $expression_value = $this->argument('value');
         $reviewed = $this->option('force');
-        $expressions = require base_path('resources/lang/'.$language.'/'.$file.'.php');
+        $path = 'resources/lang/'.$language.'/'.$file.'.php';
+        $expressions = file_exists($path) ? require base_path($path) : [];
```
Comment (kdmnk) [inline app/Console/Commands/AddExpression.php:52]: Fixed bug on locale:add artisan command causing error if the target file does not exist.

File: config/app.php [hunk 3]
```diff
@@ -90,11 +90,14 @@
...
-      'sp' => 'Spanish', 
-      'la' => 'Latina',
+        'hu' => 'Magyar',
+        'en' => 'A - English',
+        'la' => 'L - Latina',
+        'fr' => 'F - Français',
+        'it' => 'O - Italiano',
+        'de' => 'N - Deutsch',
+        'sp' => 'S - Español',
+        'gr' => 'Ó - Ελληνικά',
```
Comment (kdmnk) [inline config/app.php:100]: Enabled all the ALFONSÓ languages

File: resources/views/layouts/navbar.blade.php [hunk 4]
```diff
@@ -36,9 +36,7 @@ class="material-icons">menu</i></a>
     </div>
     <ul id="dropdownLang" class="dropdown-content">
         @foreach (config('app.locales') as $code => $name)
-        @if ($code != App::getLocale())
         <li><a href="{{ route('setlocale', $code) }}">{{ $name }}</a></li>
-        @endif
```
Comment (kdmnk) [inline resources/views/layouts/navbar.blade.php:41]: Deleted 'if' only just to see ALFONSÓ properly

**Comment by kdmnk** [comment]
And one more thing: the admin page is now accessible by every user. I don't know what would be the best solution: should I create a new role (limguist collegists could approve contributions) or a middleware/policy based on the current admin roles is enough?

**Comment by luksan47** [comment]
> And one more thing: the admin page is now accessible by every user. I don't know what would be the best solution: should I create a new role (linguist collegists could approve contributions) or a middleware/policy based on the current admin roles is enough?

I think we could create a role and give it to the collegists we trust.

File: resources/lang/hu/localizations.php [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+return [
+    'change_not_hu_en' => 'Válassz egy nyelvet, ami nem angol, vagy magyar!',
+    'help_translate' => 'Segíts nekünk fordítani!',
+    'help_translate_info' => 'Válassz egy nyelvet, amiben szeretnél segíteni fordítani. Ha látsz egy kifejezést, ami nem helyes azon a nyelven, javítsd ki, küldd el, és ha egy rendszergazda elfogadja a javításodat, akkor megjelenik az az :appban is! (Csak azok a kifejezések láthatóak, amelyekhez nem tartozik felülvizsgálatra váró változtatás.)',
```
Comment (luksan47) [inline resources/lang/hu/localizations.php:6]: nit: you could write "adminisztrátor" instead of "rendszergazda" since not only sysadmins will check these.

File: resources/views/localizations.blade.php [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+                <div class="card-title">@lang('localizations.help_translate')</div>
+                <blockquote>@lang('localizations.help_translate_info', ['app' => config('app.name')])</blockquote>
+                <div class="row">
+                    {{-- Change language --}}
+                    <div class="col s12 m3">
+                        <a class='dropdown-trigger btn' style="width: 100%" href='#' data-target='dropdownLang'>{{ config('app.locales')[App::getLocale()] }}
+                            <i class="material-icons right">arrow_drop_down</i></a>
+                        <ul id='dropdownLang' class='dropdown-content'>
+                            @foreach (config('app.locales') as $code => $name)
+                            <li><a href="{{ route('setlocale', $code) }}">{{ $name }}</a></li>
```
Comment (luksan47) [inline resources/views/localizations.blade.php:20]: We should exclude english and hungarian here.

File: resources/views/localizations.blade.php [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+            <div class="card">
+                <div class="card-content">
+                    <div class="card-title" id="{{ $fname }}">{{ $fname }}</div>
+                    @foreach ($expressions as $key => $value)
+                    {{-- Not shown if the expression have an ongoing, unapproved change --}}
+                    @php
+                    $duplicate = App\LocalizationContribution::where('language', App::getLocale())->where('key',  $fname.'.'.$key)->where('approved', false)->first();
+                    @endphp
+                    @if(is_string($value) && $duplicate == null)
+                    <form method="POST" action="{{ route('localizations.add') }}">
```
Comment (luksan47) [inline resources/views/localizations.blade.php:97]: This should be done through AJAX calls. Let's open an issue for that.

File: resources/views/localizations.blade.php [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        @php
+        $fname = substr($file, 0, -4); //filename without .php, used as a part of the expression key
+        $expressions = require base_path('resources/lang/'.$lang.'/'.$file);
+        @endphp
+        <div class="col s12">
+            <div class="card">
+                <div class="card-content">
+                    <div class="card-title" id="{{ $fname }}">{{ $fname }}</div>
+                    @foreach ($expressions as $key => $value)
+                    {{-- Not shown if the expression have an ongoing, unapproved change --}}
```
Comment (luksan47) [inline resources/views/localizations.blade.php:92]: Okay, maybe later we could find some cooler solution.

File: database/seeds/UsersTableSeeder.php [hunk 9]
```diff
@@ -66,6 +66,7 @@ private function createAdmin()
         }
         $user->roles()->attach(Role::getId(Role::PRINT_ADMIN));
         $user->roles()->attach(Role::getId(Role::INTERNET_ADMIN));
+        $user->roles()->attach(Role::getId(Role::LOCALE_ADMIN));
```
Comment (luksan47) [inline database/seeds/UsersTableSeeder.php:69]: This is enough to get this feature out. Later we should use the object field for this role so we can assign this privilege for each language separately.

File: .gitignore (hunk 1 above)
Comment (luksan47) [reply .gitignore:14]: Can we import these into the database? Maybe we could run some script?

**Review by luksan47 — CHANGES_REQUESTED** [review]
Awesome! Thanks!
Left some small requests and mostly questions and feature requests for later. Please open issues for these.

File: resources/views/admin/localizations.blade.php [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+                        @endforeach
+                    </tbody>
+                </table>
+                <div class="row">
+                    <div class="col">
+                        @if(count($contributions) == 0)
+                        Nothing to show.
+                        @else
+                        <i class="right">*You have to change to the corresponding language.</i>
+                        @endif
```
Comment (luksan47) [inline resources/views/admin/localizations.blade.php:59]: I don't get this part

File: .gitignore (hunk 1 above)
Comment (kdmnk) [reply .gitignore:14]: Hmm... Do you mean add the existing translations to this new table? Maybe. But the script have to run after addig the database and before deleting the files. And where to place the script? We would need this only once. Should we do it independently of git?

Comment (luksan47) [reply .gitignore:14]: The files won't be deleted if we pull this PR, they will be simply ignored by git. You could add it maybe as an artisan command since we already have most of the stuff there and then we will remove in a later PR.

File: resources/views/admin/localizations.blade.php (hunk 10 above)
Comment (kdmnk) [reply resources/views/admin/localizations.blade.php:59]: When you want to see the previous translations before approving the newer one, the previous is shown with @lang(...) directive (I didn't want to mess with the files, again)... So you have to change locale to the translation's language.

File: .gitignore (hunk 1 above)
Comment (kdmnk) [reply .gitignore:14]: Okay. And we will remove the files later, right?

File: resources/views/localizations.blade.php (hunk 7 above)
Comment (kdmnk) [reply resources/views/localizations.blade.php:97]: I fixed it here, UX is much better.

**Comment by kdmnk** [comment]
- Changed forms to ajax requests
- Added validation.php files for all languages
- Added locale:migrate command to add existing translations (except hu, en) to the new table
- Other minor changes, and fixes

After merging this to release, use locale:migrate and approve the translations. After this, we can delete the translation files from git (except hu, en) and this command also.

**Review by luksan47 — CHANGES_REQUESTED** [review]
Awesome!
One small problem, now the old and new value are always the same on the locale admin page. Could you check it?

**Comment by kdmnk** [comment]
> Awesome!
> One small problem, now the old and new value are always the same on the locale admin page. Could you check it?

Thanks, I fixed it

**Comment by luksan47** [comment]
> > Awesome!
> > One small problem, now the old and new value are always the same on the locale admin page. Could you check it?
> 
> Thanks, I fixed it

It's still not working for me, now the new values is constantly "en".

**Comment by kdmnk** [comment]
> > > Awesome!
> > > One small problem, now the old and new value are always the same on the locale admin page. Could you check it?
> > 
> > 
> > Thanks, I fixed it
> 
> It's still not working for me, now the new values is constantly "en".

Have you ran npm run dev, and deleted the cache?

**Comment by luksan47** [comment]
> > It's still not working for me, now the new values is constantly "en".
> 
> Have you ran npm run dev, and deleted the cache? @luksan47

My bad, it works. Thanks!

**Review by luksan47 — APPROVED** [review]

**Comment by luksan47** [comment]
Please open the followup issues.
